Add get_embedding_dofs() to FE_Nedelec (2-D)

harmonj opened the pull request:
Adds `get_embedding_dofs()` to the `FE_Nedelec` element (currently for `dim==2`).

To make this function easily accessible for `FE_Nedelec` elements and other hierarchical elements (like `FE_Q_Hierarchical`), the `FiniteElement` class now includes an associated virtual function.

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+            dofs_sup[i] *
+            fe_sup.shape_value_component(i, quadrature.point(k), comp);
+
+        double diff = std::abs(eval_sub - eval_sup);
```
```suggestion
        const double diff = std::abs(eval_sub - eval_sup);
```

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+
+  // Assign arbitrary values to dofs_sub
+  for (unsigned int i = 0; i < dofs_sub.size(); ++i)
+    dofs_sub[i] = sin(i + 0.5);
```
```suggestion
    dofs_sub[i] = std::sin(i + 0.5);
```
Only cosmetic change. I'd suggest to be more specific here.

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+  // Set the quadrature sufficiently high for the enriched finite element space
+  QGauss<dim> quadrature(fe_sup.degree + 1);
+
+  std::vector<double> dofs_sub(fe_sub.dofs_per_cell),
+    dofs_sup(fe_sup.dofs_per_cell);
```
```suggestion
  std::vector<double> dofs_sub(fe_sub.dofs_per_cell);
  std::vector<double> dofs_sup(fe_sup.dofs_per_cell);
```
We applied this practice elsewhere to increase readability.

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+  Assert((sub_degree > 0) && (sub_degree <= sup_degree),
+         ExcIndexRange(sub_degree, 1, sup_degree));
+
+  FE_Nedelec<dim> fe_sub(sub_degree - 1), fe_sup(sup_degree - 1);
```
```suggestion
  const FE_Nedelec<dim> fe_sub(sub_degree - 1);
  const FE_Nedelec<dim> fe_sup(sup_degree - 1);
```

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+          << fe_sub.dofs_per_cell << ' ' << fe_sup.dofs_per_cell;
+
+  // Set the quadrature sufficiently high for the enriched finite element space
+  QGauss<dim> quadrature(fe_sup.degree + 1);
```
```suggestion
  const QGauss<dim> quadrature(fe_sup.degree + 1);
```

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+
+
+  // Map the DoFs from fe_sub to fe_sup
+  std::vector<unsigned int> embedding_dofs =
```
```suggestion
  const std::vector<unsigned int> embedding_dofs =
```

marcfehling commented on tests/fe/nedelec_embed_dofs.cc:
```diff
@@ -0,0 +1,90 @@
+
+#include "interpolate_common.h"
+
+template <unsigned int dim>
```
```suggestion
template <int dim>
```
You are right technically. Throughout the library however, we use just `int dim` for template parameters, so I suggest to do so here as well.

marcfehling commented on include/deal.II/fe/fe.h:
```diff
@@ -2275,6 +2275,16 @@ class FiniteElement : public Subscriptor, public FiniteElementData<dim>
+  /**
+   * For a finite element of degree @p sub_degree < @p degree, we return a
+   * vector which maps the numbering on an FE of degree @p sub_degree into the
+   * numbering on this element. This function is appropriate for hierarchical
```
```suggestion
   * numbering on this element.
   *
   * This function is appropriate for hierarchical
```
Maybe separate this into two paragraphs.

marcfehling commented on include/deal.II/fe/fe_nedelec.h:
```diff
@@ -324,6 +324,16 @@ class FE_Nedelec : public FE_PolyTensor<dim>
+  /**
+   * For a finite element of degree @p sub_degree < @p degree, we return a
+   * vector which maps the numbering on an FE of degree @p sub_degree into the
+   * numbering on this element. Note that for the Nedelec element, by @p sub_degree,
```
```suggestion
   * numbering on this element.
   * 
   * Note that for the Nedelec element, by @p sub_degree,
```
Same here.

marcfehling commented on source/fe/fe_nedelec.cc:
```diff
@@ -4121,7 +4121,61 @@ FE_Nedelec<dim>::memory_consumption() const
+      case 2:
+        {
+          // The Nedelec cell has only Face (Line) and Cell DoFs...
+          unsigned int n_face_dofs_sub =
```
```suggestion
          const unsigned int n_face_dofs_sub =
```

marcfehling commented on source/fe/fe_nedelec.cc:
```diff
@@ -4121,7 +4121,61 @@ FE_Nedelec<dim>::memory_consumption() const
+          // The Nedelec cell has only Face (Line) and Cell DoFs...
+          unsigned int n_face_dofs_sub =
+            GeometryInfo<dim>::lines_per_cell * sub_degree;
+          unsigned int n_cell_dofs_sub = 2 * (sub_degree - 1) * sub_degree;
```
```suggestion
          const unsigned int n_cell_dofs_sub = 2 * (sub_degree - 1) * sub_degree;
```

marcfehling commented on source/fe/fe_nedelec.cc:
```diff
@@ -4121,7 +4121,61 @@ FE_Nedelec<dim>::memory_consumption() const
+          // Identify the Face/Line DoFs
+          while (i < n_face_dofs_sub)
+            {
+              unsigned int face_index = i / sub_degree;
```
```suggestion
              const unsigned int face_index = i / sub_degree;
```

marcfehling commented on source/fe/fe_nedelec.cc:
```diff
@@ -4121,7 +4121,61 @@ FE_Nedelec<dim>::memory_consumption() const
+          // Identify the Cell DoFs
+          if (sub_degree >= 2)
+            {
+              unsigned int n_dofs_faces =
```
```suggestion
              const unsigned int n_face_dofs =
```
You called a variable above `n_face_dofs_sub`, so I would suggest to stick to that naming scheme.

marcfehling commented on source/fe/fe_nedelec.cc:
```diff
@@ -4121,7 +4121,61 @@ FE_Nedelec<dim>::memory_consumption() const
+                                        this->degree * (this->degree - 1) +
+                                        (ku - 2) * (this->degree) + kv;
+            }
+          return embedding_dofs;
```
Before returning the result, is there some way you can verify that all entries are set? Maybe something like
```suggestion
          Assert(i == (n_face_dofs_sub + n_cell_dofs_sub), ExcInternalError());
          return embedding_dofs;
```

harmonj replied:
Yes, that works. I added this (along with all your other suggestions above) in the new commit.

marcfehling reviewed (changes requested):
Code looks great! I have just some minor suggestions, mainly turning variables into `const`.

harmonj commented:
Thank you, @marcfehling!

I added each of your suggestions above and squashed the commits. Please let me know if you'd like to see other changes before merging.

marcfehling commented:
/rebuild

bangerth commented on include/deal.II/fe/fe_nedelec.h:
```diff
@@ -324,6 +324,18 @@ class FE_Nedelec : public FE_PolyTensor<dim>
+   * opposed to the Nedelec degree.
+   */
+  virtual std::vector<unsigned int>
+  get_embedding_dofs(const unsigned int sub_degree) const override;
```
Even for the Nedelec element, isn't this only applicable for, say, element orders 1 and n, 2 and 2n, 3 and 3n (assuming we are using equispaced points), etc?

harmonj replied:
Perhaps I have misunderstood portions of the Nedelec element implementation in deal.II, but if this were the case I would have expected that this would have been indicated by the test I wrote. I also tested beforehand with the original way I was mapping from p-reduced Nedelec cells to enriched ones using the curl-conforming projection-based interpolation.

I'll have to dive back into the `FE_Nedelec` implementation and see if there is something I missed that might prevent identification of equivalent dofs between a Nedelec cell of `sub_degree` and a Nedelec cell of `sup_degree` in this manner for some cases.

bangerth replied:
Think about the usual Q_k elements in 2d. If you have a Q_2, it has DoFs at the vertices, edge midpoints, and cell midpoint. If you have a Q_3, it has DoFs at the vertices but also at positions 1/3 and 2/3 along the edge, and 4 in the interior of the cell which are all not located at the cell center. I don't think the Nedelec element has a structure that is different from the general problem outlined above unless it is *moment based* along edges in which all edge and cell degrees of freedom might be located at edge and cell midpoints.

harmonj replied:
The case for `FE_Nedelec` is not quite the same as the standard Q_k element. The support points are the Gauss-Legendre quadrature points (on the faces or the interior of the cell), but the identification of equivalent DoFs from a Nedelec element of `sub_degree` to an element of `sup_degree` is based on the hierarchical construction of the mixed order Nedelec polynomials as in `PolynomialsNedelec<dim>::create_polynomials()`. In other words, the enriched cell includes all the shape functions of the reduced cell, so there is no concern w.r.t. the choice of `sub_degree` or `sup_degree`.

bangerth reviewed (commented):
I don't think this function should be in the base class. But that shouldn't mean that it can't be in the Nedelec class itself, of course.

bangerth reviewed (dismissed):
I don't think this function should be in the base class. But that shouldn't mean that it can't be in the Nedelec class itself, of course.

harmonj commented on include/deal.II/fe/fe.h:
```diff
@@ -2275,6 +2275,18 @@ class FiniteElement : public Subscriptor, public FiniteElementData<dim>
+   * therefore, may not implement it.
+   */
+  virtual std::vector<unsigned int>
+  get_embedding_dofs(const unsigned int sub_degree) const;
```
Yes, since the one to one identification is only possible in special cases I agree that it makes more sense to leave this out of the base class.

I'll take a look at implementing `get_interpolation_matrix()` for a better generalization here.

harmonj commented:
I removed the function from the base class for the reasons stated above and squashed the commits.

Somewhat unrelated to this specific commit but related to the identification of embedding DoFs, would it be of interest to add illustrations of the shape functions (by component) on the reference cell for `FE_Nedelec` in the documentation as seen for other finite element classes?

bangerth commented:
More documentation is always better! It's more difficult for vector-valued elements, of course, but could probably also be scripted. I wished I still knew where the script is (or whether it was preserved at all) with which we generated the existing figures. I know that it used a mesh with a single cell, a solution vector with exactly one one in it, and `DataOut::build_patches()` with a large number of subdivisions.

But if you wanted to go down that route, let's do it in a separate PR.

harmonj commented:
> More documentation is always better! It's more difficult for vector-valued elements, of course, but could probably also be scripted. I wished I still knew where the script is (or whether it was preserved at all) with which we generated the existing figures. I know that it used a mesh with a single cell, a solution vector with exactly one one in it, and `DataOut::build_patches()` with a large number of subdivisions.


Sounds good. I'll have to play around with the visualization for the script I wrote last week to plot these shape functions and keep things consistent with the other elements.  I'll do that for a separate PR.

jppelteret commented:
> Somewhat unrelated to this specific commit but related to the identification of embedding DoFs, would it be of interest to add illustrations of the shape functions (by component) on the reference cell for FE_Nedelec in the documentation as seen for other finite element classes?

If it helps, I placed a small program in https://github.com/dealii/dealii/issues/8634#issuecomment-595715677 that outputs all shape functions in the VTU format. It works for the Nedelec elements too.

harmonj commented:
> If it helps, I placed a small program in [#8634 (comment)](https://github.com/dealii/dealii/issues/8634#issuecomment-595715677) that outputs all shape functions in the VTU format. It works for the Nedelec elements too.

Thank you. The script that I wrote is similar, though I was plotting the components individually as scalar quantities. I will experiment with the visualization and submit a separate pull request soon.

Otherwise, this PR should be good to go, unless there are other changes anyone would like me to make.

marcfehling commented on include/deal.II/fe/fe_nedelec.h:
```diff
@@ -324,6 +324,18 @@ class FE_Nedelec : public FE_PolyTensor<dim>
   clone() const override;
 
+  /**
+   * For a finite element of degree @p sub_degree < @p degree, we return a
```
```suggestion
   * For a finite element of degree larger than @p sub_degree, we return a
```
I think this is easier to read.

harmonj replied:
I agree that this is easier to understand. I pushed this change.

Would you like me to make the same modification to the description for the equivalent function in `FE_Q_Hierarchical`?

marcfehling replied:
> Would you like me to make the same modification to the description for the equivalent function in `FE_Q_Hierarchical`?

I think that this would be a nice improvement, so go for it! I suggest opening a separate PR for this change.

harmonj replied:
Sounds good. Will do.

peterrum commented:
Is this PR ready to be merged?

harmonj commented:
> Is this PR ready to be merged?

Yes, I believe so.

kronbichler reviewed (approved):
I think we somehow forgot to merge this one. @tamiko do we want to still merge it for the upcoming release, as the milestone suggests? I would be in favor, but we need a second opinion at this stage.

tamiko commented:
We need in particular @bangerth to OK the pull request :-)

bangerth commented:
My apologies for dropping the ball in the first place (and then being on vacation until Monday and in the air Tuesday+Wednesday). Thanks everyone for handling the patch, and to @harmonj for writing it in the first place!
